based/bin/portal/src/clients/gateway.rs

```diff
@@ -0,0 +1,486 @@
+use bop_common::{
+    api::{BasedAuthApiClient, ControlApiClient, EngineApiClient, OpMinerExtApiClient, RegistryApiClient},
+    auth::gateway_auth_message,
+    communication::Producer,
+    debug_panic,
+    metrics::{Gauge, Metric, MetricsUpdate},
+    signing::ECDSASigner,
+    time::{Duration, Instant},
+    utils::uuid,
+};
+use eyre::eyre;
+use indexmap::IndexMap;
+use jsonrpsee::{
+    core::{ClientError, middleware::layer::RpcLogger},
+    http_client::{HttpClient, HttpClientBuilder, HttpRequest, RpcService, transport::HttpBackend},
+};
+use op_alloy_rpc_types_engine::{OpExecutionPayloadV4, OpPayloadAttributes};
+use reqwest::{
+    Url,
+    header::{AUTHORIZATION, HeaderValue, InvalidHeaderValue},
+};
+use tokio::sync::RwLock;
+use tower::{Layer, Service};
+use tracing::{Instrument, debug, error, info, trace, warn};
+
+use super::RpcClient;
+use crate::{cli::PortalArgs, clients::create_client};
+
+pub type GatewayClient = HttpClient<RpcLogger<RpcService<GatewayClientService<HttpBackend>>>>;
+
+#[derive(Debug)]
+pub struct GatewayClientAuthLayer {
+    pub token: Arc<HeaderValue>,
+}
+impl GatewayClientAuthLayer {
+    pub fn new(token: &str) -> Result<Self, InvalidHeaderValue> {
+        let value = HeaderValue::from_str(token)?;
+        Ok(Self { token: Arc::new(value) })
+    }
+}
+
+impl<S> Layer<S> for GatewayClientAuthLayer {
+    type Service = GatewayClientService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        GatewayClientService { jwt: self.token.clone(), inner }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GatewayClientService<S> {
+    jwt: Arc<HeaderValue>,
+    inner: S,
+}
+
+impl<S, B> Service<HttpRequest<B>> for GatewayClientService<S>
+where
+    S: Service<HttpRequest<B>>,
+{
+    type Error = S::Error;
+    type Future = S::Future;
+    type Response = S::Response;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: HttpRequest<B>) -> Self::Future {
+        req.headers_mut().insert(AUTHORIZATION, Arc::unwrap_or_clone(self.jwt.clone()));
+        self.inner.call(req)
+    }
+}
+
+#[derive(Clone)]
+pub struct Gateway {
+    pub url: Url,
+    pub address: Arc<Address>,
+    pub client: Arc<RwLock<Option<GatewayClient>>>,
+    pub ping: Arc<RwLock<Duration>>,
+    pub active: Arc<AtomicBool>,
+    pub registry_index: Arc<AtomicU64>,
+    pub metrics: Producer<MetricsUpdate>,
+}
+
+impl Gateway {
+    pub async fn health_check(&self) {
+        let Some(client) = self.client.read().await.clone() else {
+            self.active.store(false, Ordering::Relaxed);
+            return;
+        };
+        let ping_start = Instant::now();
+        match ControlApiClient::heartbeat(&client).await {
+            Ok(_) => {
+                let ping_duration = ping_start.elapsed();
+                *self.ping.write().await = ping_duration;
```

Comment:
Why use an async RwLock here?

Reply:
Changed to an arc cell (and also some other locks surrounding these items)